psychopy/visual/textbox2/textbox2.py

```diff
@@ -559,63 +562,71 @@ def alignment(self, alignment):
 
     @property
     def text(self):
-        lastFormatter = NONE
-        formatted_text = ''
-        styles = self._styles
-        for i, c in enumerate(self._text):
-            if styles[i] == ITALIC and lastFormatter != styles[i]:
-                formatted_text+='<i>%s'%(c)
-            elif styles[i] == BOLD and lastFormatter != styles[i]:
-                formatted_text+='<b>%s'%(c)
-            
-            elif styles[i] != ITALIC and lastFormatter == ITALIC:
-                formatted_text+='</i>%s'%(c)
-            elif styles[i] != BOLD and lastFormatter == BOLD:
-                formatted_text+='</b>%s'%(c)
-            else:
-                formatted_text+=c
-            lastFormatter = styles[i]
-        return formatted_text
+        return self._styles.formatted_text
     
     @text.setter
     def text(self, text):
         # Convert to string
         text = str(text)
+        original_text = text
         # Substitute HTML tags
         text = text.replace('<i>', codes['ITAL_START'])
         text = text.replace('</i>', codes['ITAL_END'])
         text = text.replace('<b>', codes['BOLD_START'])
-        text = text.replace('</b>', codes['BOLD_END'])      
+        text = text.replace('</b>', codes['BOLD_END'])
+        text = text.replace('</c>', codes['COLOR_END'])
+
+        # Handle starting color tag
+        color_values = re.findall(re_hex_pattern, text)
+
+        # Only execute if color codes are found to save a regex call
+        if len(color_values) > 0:
+            text = re.sub(re_hex_pattern, codes['COLOR_START'], text)
+            
+            # Convert color_values from markdown+hex to RGBA (0, 1) for _layout
+            color_values = [Color(v[3:-1], 'hex').render('rgba1') for v in color_values]
```

Comment:
I think it's likely that this is the source of the slower times you found in your tests - once a Color object is 'render'ed, it caches the value so that future render calls in the same space just return the already calculated value (until an attribute of the Color object has changed). This means you can do:
```
col = Color("white", 'rgb')
for n in range(1000):
    col.render("rgba1")
```
and the time taken will just be first calculation time + the time taken to index an array * 1000, rather than calculation time * 1000. Creating a new Color object means creating a new cache, so:
```
for n in range(1000):
    col = Color("white", 'rgb')
    col.render("rgba1")
```
will be a lot slower. I have an idea for how to do this more efficiently so I'll put in a PR to your branch :)

Reply:
Thanks for looking at the PR and for your suggestion!

I did originally have a different solution, but didn't find that it was any faster/slower during the aforementioned test:

```
color_values_ = []
for v in color_values:
    h = v[3:-1] # hex string (with or without #)
    h = h.strip('#')
    
    # Map from hex -> RGB (0,255) -> RGBA (0, 1). Alpha always == 1.0
    r = int(h[0:2], 16) / 255.
    g = int(h[2:4], 16) / 255.
    b = int(h[4:6], 16) / 255.
    a = 1. # value doesn't seem to affect rendering in _layout
    color_values_.append((r, g, b, a))
color_values = color_values_
```

Although it's possible that even this is slow (although I did try to optimize this by getting rid of any unnecessary loops)

Reply:
I don't think I have the permissions to submit a PR to your branch, but this is what I was thinking: https://github.com/TEParsons/psychopy/tree/textbox2_colors

So once you've got a color specified once in text, you cache the object handle so that when the same color is specified again you're getting the rgba value from the same object, meaning it won't calculate it again

Reply:
I've added you as a direct collaborator, hopefully that fixes it. I can also manually implement your changes if that's more convenient, just let me know (although I will be heading off to bed shortly since it's 2:30)

Your solution makes sense to me and it looks good! I can test/merge it when I wake up later today

Reply:
That would be fab, thanks! I've submitted a PR now :)